lib/bridge/matrix-handler.js

```diff
@@ -568,25 +584,78 @@ let _onAliasQuery = Promise.coroutine(function*(req, roomAlias) {
     );
 
     if (matrixRooms.length === 0) {
+        // ====== Track the IRC channel
         // lower case the name to join (there's a bug in the IRC lib
         // where the join callback never fires if you try to join
         // #WithCaps in channels :/)
         channelInfo.channel = toIrcLowerCase(channelInfo.channel);
-        req.log.info("Going to track channel %s", channelInfo.channel);
+        req.log.info("Going to track IRC channel %s", channelInfo.channel);
         // join the irc server + channel
         yield ircLib.trackChannel(channelInfo.server, channelInfo.channel);
-        req.log.info("Bot is now tracking channel.");
-        let matrixRoom = yield req.mxLib.createRoomWithAlias(
-            roomAlias, channelInfo.channel, undefined,
-            channelInfo.server.getJoinRule(),
-            channelInfo.server.shouldFederate(),
-            channelInfo.server.shouldPublishRooms()
-        );
+        req.log.info("Bot is now tracking IRC channel.");
+
+        // ======== Create the Matrix room
+        let newRoomId = null;
+        let botIntent = this.ircBridge.bridge.getIntent();
+        try { // make the matrix room
+            let visibility = undefined;
+            let shouldSetJoinRule = true;
+            if (channelInfo.server.shouldPublishRooms()) {
+                visibility = "public";
+                if (channelInfo.server.getJoinRule() === "public") {
+                    shouldSetJoinRule = false; // visibility=public also sets join_rule=public
+                }
+            }
+            else {
+                visibility = "private";
+                if (channelInfo.server.getJoinRule() === "invite") {
+                    shouldSetJoinRule = false; // visibility=private also sets join_rule=invite
+                }
+            }
+
+            let res = yield botIntent.createRoom({
+                options: {
+                    room_alias_name: roomAlias.split(":")[0].substring(1), // localpart
+                    name: channelInfo.channel,
+                    topic: undefined,
+                    visibility: visibility,
+                    creation_content: {
+                        "m.federate": channelInfo.server.shouldFederate()
+                    }
+                }
+            });
+
+            newRoomId = res.room_id;
+
+            if (shouldSetJoinRule) {
+                yield botIntent.sendStateEvent(newRoomId, "m.room.join_rules", "", {
+                    join_rule: channelInfo.server.getJoinRule()
+                });
+            }
+
+            yield botIntent.sendStateEvent(newRoomId, "m.room.history_visibility", "", {
+                history_visibility: "joined"
+            });
+        }
+        catch (e) {
+            if (e && e.errorcode === "M_UNKNOWN") {
+                // alias already taken, must be us. Join the room alias.
```

Comment:
The intent pattern seems to be to try to join, and then do the creation if it failed. Why is this the other way around?